protected-api-access-protected-api/README.md

```diff
@@ -0,0 +1,145 @@
+- Node.Js 16+
+- An existing Web API
+
+## Setup
+
+### 1. Modify your existing Web API
+
+| App registration <br/> setting    | Value for this sample app                                                    | Notes                                                                                              |
+|:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
+| **Expose an API scope**           | api://{clientid}/user_impersonation                                          | Required value for this sample. <br/> The Node Web App registration will reference this value.     |
+| **accessTokenAcceptedVersion**    | 2                                                                            | Required value for this sample.                                                                    |
+
+### 2. Register Node Web App
+
+Complete the steps in [Register an application with the Microsoft identity platform](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app) to register the sample app.
+
+Use these settings in your app registration.
+
+| App registration <br/> setting    | Value for this sample app                                                    | Notes                                                                                              |
+|:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
+| **Name**                          | `Node Web APP`                                                               | Suggested value for this sample. <br/> You can change the app name at any time.                    |
+| **Supported account types**       | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
+| **Platform type**                 | **Web**                                                                      | Required value for this sample.                                                                    |
+| **Redirect URI**                  | `http://localhost:8080/redirect`                                             | Required value for this sample.                                                                    |
+| **Client secret**                 | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |
+| **API Permissions**               | `user_impersonation`                                                         | Create a new delegated permission called user_impersonation.  Required value for this sample.      |
```

Comment:
This is the on-behalf-of flow, so it's pretty common to see the `user_impersonation` string being used as the scope here, as it's not asking for any additional scopes, it's simply expecting the access token to be sufficient for the upstream call.  A lot of our examples do so it that way, as do a lot of Azure services as well, it's kinda a "thing" at this point, but honestly, I'm not a huge fan because that string almost looks like it's an "important" string (naming wise).  Conventions are a funny thing.

Reply:
OK, we'll roll with `user_impersonation` for now because this needs straightening out elsewhere.

The reason I asked this is because we have an open issue from a customer asking scope naming about convention:

> **Documentation says naming conventation is both access_as_user and Employees.Read.All**
> https://github.com/MicrosoftDocs/azure-docs/issues/84351

@jmprieur recommends `access_as_user` in https://github.com/MicrosoftDocs/azure-docs/issues/68540#issuecomment-756112758). It's also what we have in [Protected web API: App registration](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-protected-web-api-app-registration#exposing-delegated-permissions-scopes).

However, as you see in the quickstart, we use `resource.operation.constraint`.
 
None of them, however, specifies when to use which scope naming convention. I might sync up with @hpsin and see if we can get the story straight so I can update docs as needed.

Reply:
resource.operation.constraint is correct.  It just gets ignored because, as @ckittel points out, people have been ignoring it for a long time (including, of course, the defaults in Azure Portal). 

In Azure this is common because they use RBAC for authorization, and the operations, resources, and constraints are held there, in an expressive policy that supports delegation restrictions.  If Office were to use this we'd be in trouble.

Reply:
> If Office were to use **this** we'd be in trouble.

@hpsin Which *this* are you referring to here? :)

Reply:
If Office just used user_impersonation, or Graph did we'd be in trouble, because e.g. every user has permission to read their own email.  If Graph had just the one scope, you'd be unable to indicate "app Foo is allowed to read my email but not edit my files".  

Azure doesn't have the same conception of resources - you don't "have" files in Azure that are yours alone. They exist, and have their own ACLs, which extend to the user and to the app calling.